numba/core/debuginfo.py

```diff
@@ -315,8 +315,11 @@ def mark_subprogram(self, function, qualname, argnames, argtypes, line):
                                        line=line, function=function,
                                        argmap=argmap)
         function.set_metadata("dbg", di_subp)
-        # disable inlining for this function for easier debugging
-        function.attributes.add('noinline')
+
+        # Don't marked alwaysinline functions as noinline.
+        if 'alwaysinline' not in function.attributes:
+            # disable inlining for this function for easier debugging
+            function.attributes.add('noinline')
```

Comment:
Why not the other way around? Discard `alwaysinline` is `noinline` is present. The `noinline` is being added for better debugging experience.

Reply:
I left it this way around because it matches up with the "direct lowering" experience that is present in Numba < 0.54, and also will be in 0.55. I guess we can swap it but it might get weird as e.g. potentially trivial things like `operator.add` would end up as a `noinline` function call, this making the IR pretty strange looking?

Reply:
hm... i'll need to play around with the other way.

Reply:
With everything as `noinline`, I don't find it strange, but retaining the user specified `forceinline` seems more intuitive and give users the control.

With code like:

```python
import math
import numpy as np
from numba import njit

@njit(forceinline=True)
def bar(x):
    return math.sin(x)

@njit(forceinline=False)
def baz(x):
    return math.cos(x)

@njit(forceinline=True)
def array_work(a, b):
    c = a + np.zeros(10) * b
    return np.sum(c)

@njit
def foo(x):
    # gdb()
    a = bar(x)
    b = baz(x)
    return array_work(a, b)

result = foo(np.pi)
foo.inspect_cfg(foo.signatures[0]).display(view=True)
```

If everything is `noinline`, it will show:

<img width="211" alt="Screen Shot 2021-12-10 at 5 34 05 PM" src="https://user-images.githubusercontent.com/1929845/145654169-c6da2e3d-2179-4698-b962-364b1e4336e8.png">

If we use the changes in this PR, it will show:

<img width="211" alt="Screen Shot 2021-12-10 at 5 34 35 PM" src="https://user-images.githubusercontent.com/1929845/145654170-38f8e649-ab68-4182-94a0-83b092b8dc18.png">

If I'm debugging, both versions are useful.